File: 0031-next-permutation/0031-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size(), MAX = nums[nums.size() - 1];
        multiset<int> mini;
        for (int i = n - 1; i >= 0; --i) {
            mini.insert(nums[i]);
            if (nums[i] < MAX) {
                auto t = mini.upper_bound(nums[i]);
                nums[i] = *t;
                mini.erase(t);
                for (auto j : mini) {
                    nums[++i] = j;
                }
                return;
            }
            MAX = max(MAX, nums[i]);
        }
        int i = 0;
        for (auto j : mini) {
            nums[i++] = j;
        }
    }
};

File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        Integer pos = nums.length - 2;
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Comparator.reverseOrder());
        maxHeap.add(nums[nums.length - 1]);
        while(pos >= 0) {
            if (nums[pos] < maxHeap.peek()) {
                int top = maxHeap.peek(), temp = nums.length - 1;
                maxHeap.poll();
                while(maxHeap.size() > 0 && maxHeap.peek() > nums[pos]) {
                    nums[temp--] = top;
                    top = maxHeap.poll();
                }
                nums[temp--] = nums[pos];
                nums[pos] = top;
                while(maxHeap.size() > 0) {
                    nums[temp--] = maxHeap.poll();
                }
                return;
            }
            maxHeap.add(nums[pos]);
            --pos;
        }
        Arrays.sort(nums);
    }
}